File: core/src/main/java/org/loopring/lightcone/core/actors/BalanceCacher.java
/*
 * Copyright 2018 Loopring Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.loopring.lightcone.core.actors;

import akka.actor.*;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import org.loopring.lightcone.proto.Balance;
import org.loopring.lightcone.proto.Cache;
import org.loopring.lightcone.proto.Cache.CachedAddressBalanceInfo;
import org.loopring.lightcone.proto.Cache.CacheAddressBalanceInfo;

import java.util.Optional;

public class BalanceCacher extends AbstractActor {
    private final LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);

    public static Props props(Optional<String> settingsId) {
        return Props.create(BalanceCacher.class);
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(CacheAddressBalanceInfo.class, r -> getSender().tell(CachedAddressBalanceInfo.getDefaultInstance(), getSender()))
                .match(Cache.GetAddressBalanceFromCache.class, r -> getSender().tell(Balance.AddressBalanceInfo.getDefaultInstance(), getSender()))
                .build();
    }
}

File: core/src/main/scala/org/loopring/lightcone/core/actors/BalanceCacher.scala
/*
 * Copyright 2018 lightcore-relay
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.loopring.lightcone.core.actors

import akka.actor._
import akka.cluster.pubsub.DistributedPubSub
import akka.cluster.pubsub.DistributedPubSubMediator.Subscribe
import akka.pattern.pipe
import akka.util.Timeout
import com.google.inject.Inject
import org.loopring.lightcone.core.cache._
import org.loopring.lightcone.proto.balance._
import org.loopring.lightcone.proto.cache._
import org.loopring.lightcone.proto.deployment._

import scala.concurrent.ExecutionContext

object BalanceCacher
  extends base.Deployable[BalanceCacherSettings] {
  val name = "balance_cacher"

  def getCommon(s: BalanceCacherSettings) =
    base.CommonSettings(None, s.roles, s.instances)
}

class BalanceCacher @Inject() (cache: BalanceCache)(implicit
    ec: ExecutionContext,
    timeout: Timeout
)
  extends Actor {

  var settings: BalanceCacherSettings = null

  DistributedPubSub(context.system).mediator ! Subscribe(CacheObsoleter.name, self)

  def receive: Receive = {
    case settings: BalanceCacherSettings ⇒
      this.settings = settings
    case m: GetBalancesReq ⇒
      cache.getBalances(m).pipeTo(sender)
    case m: GetAllowancesReq ⇒
      cache.getAllowances(m).pipeTo(sender)
    case m: GetBalanceAndAllowanceReq ⇒
      cache.getBalanceAndAllowances(m).pipeTo(sender)
    case m: CacheBalanceInfo ⇒
      cache.addCache(m).pipeTo(sender)
    case purgeEvent: Purge.Balance         ⇒

    case purgeEvent: Purge.AllForAddresses ⇒

    case purgeEvent: Purge.AllAfterBlock   ⇒

    case purgeEvent: Purge.All             ⇒
  }
}
